### PR TITLE
HDDS-4838. Deleting at bucket root throws NPE.

### PR DESCRIPTION
**Pull request by sadanand48**
## What changes were proposed in this pull request?
In case of deleting the  root, In `o3fs`, the path would be pointing to a bucket, If it is not a fully qualified path, it would just be relative to the root and the path would be = "/"
path ="/"; 
NPE is obtained while doing path.getParent() as "/" parent will be null. 
```java 
String qpath = this.fs.makeQualified(path).toString();
Path trashRoot = this.fs.getTrashRoot(path);
Path trashCurrent = new Path(trashRoot, CURRENT);
Path trashPath = this.makeTrashRelativePath(trashCurrent, path);
Path baseTrashPath = this.makeTrashRelativePath(trashCurrent, path.getParent());

```
The fix here would be  would be to make getTrashRoot() return a qualified path which even is the same behaviour in HDFS. If that is fixed we would get a proper warning while deleting the root as the below condition will pass

```java 
else if (trashRoot.getParent().toString().startsWith(qpath)) {
        throw new IOException("Cannot move \"" + path + "\" to the trash, as it contains the trash");
      } 
```

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-4838

## How was this patch tested?
Unit tests

### CODE REVIEW

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFileSystem.java [hunk 1]
```diff
@@ -1638,6 +1639,26 @@ public void testRenameToTrashEnabled() throws Exception {
...
     Assert.assertTrue(o3fs.exists(trashPath));
+
+    // Try to delete root
+    Path root = new Path(OZONE_URI_DELIMITER);
+    try {
+      trash.moveToTrash(root);
+      Assert.fail("Trying to delete root should throw exception");
+    } catch (Exception e) {
+      Assert.assertTrue(e instanceof IOException);
+    }
```
Comment (ayushtkn) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFileSystem.java:1650]: Does this work:
```
    Assert.assertThrows(IOException.class, () -> trash.moveToTrash(root));
```

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFileSystem.java [hunk 2]
```diff
@@ -1638,6 +1639,26 @@ public void testRenameToTrashEnabled() throws Exception {
...
+    OzoneConfiguration conf2 = new OzoneConfiguration(cluster.getConf());
+    conf2.setClass("fs.trash.classname", TrashPolicyDefault.class,
+        TrashPolicy.class);
+    Trash trashPolicyDefault = new Trash(conf2);
+    try {
+      trashPolicyDefault.moveToTrash(root);
+      Assert.fail("Trying to delete root should throw exception");
+    } catch (Exception e) {
+      Assert.assertTrue(e instanceof IOException);
+    }
```
Comment (ayushtkn) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFileSystem.java:1661]:
```
    Assert.assertThrows(IOException.class, () -> trashPolicyDefault.moveToTrash(root));
```

Comment (sadanand48) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFileSystem.java]: Done.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFileSystem.java (hunk 1 above)
Comment (sadanand48) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestOzoneFileSystem.java]: Done.

**Review by ayushtkn — APPROVED** [review]
LGTM

**Review by ashishkumar50 — APPROVED** [review]
@sadanand48 thanks for working on this, LGTM +1.
`testRenameToTrashEnabled` is still marked as flaky, I think it will get resolve with other PR which is currently open.

**Review by aryangupta1998 — APPROVED** [review]
Thanks @sadanand48 for working on this, LGTM +1

**Comment by aryangupta1998** [comment]
Thanks, @ayushtkn and @ashishkumar50 for the reviews.

**Pull request merged**